File: agents/alphchemy/src/mcp_server/notebook_tools.py
from __future__ import annotations
from typing import Any
from supabase import Client
from analysis.format_analysis import format_query_results, format_number
from analysis.query import Query

def validate_notebook_parts(queries: list[Any], notes: list[str]) -> None:
    if len(queries) != len(notes):
        raise ValueError("queries and notes must have the same length")

def clear_query_results(queries: list[dict[str, Any]]) -> list[dict[str, Any]]:
    cleared = []

    for query in queries:
        new_query = query.copy()
        new_query["results"] = None
        cleared.append(new_query)

    return cleared

def fetch_notebook_row(supabase: Client, notebook_id: int, user_id: str) -> dict[str, Any]:
    table = supabase.table("notebooks")
    selected = table.select("id, last_updated, title, queries, notes, status, error_message")
    owned = selected.eq("user_id", user_id)
    filtered = owned.eq("id", notebook_id)
    limited = filtered.limit(1)
    rows = limited.execute().data

    if len(rows) == 0:
        raise ValueError(f"notebook id={notebook_id} not found")

    return rows[0]

def query_experiments_tool(supabase: Client, query: str, user_id: str) -> str:
    parsed = Query(query = query)
    parsed.run(supabase, user_id)
    return format_query_results(parsed)

def list_notebooks_tool(supabase: Client, user_id: str) -> str:
    table = supabase.table("notebooks")
    selected = table.select("id, last_updated, title")
    owned = selected.eq("user_id", user_id)
    ordered = owned.order("last_updated", desc = True)
    rows = ordered.execute().data

    lines = [f"[NOTEBOOKS] {len(rows)} notebook(s)"]
    for row in rows:
        lines.append(f"id={row['id']} title={row['title']}")

    return "\n".join(lines)

def view_notebook_tool(supabase: Client, notebook_id: int, user_id: str) -> str:
    row = fetch_notebook_row(supabase, notebook_id, user_id)
    lines = [
        f"id: {row['id']}",
        f"last_updated: {row['last_updated']}",
        f"title: {row['title']}",
        f"status: {row['status']}"
    ]

    error_message = row["error_message"]
    if error_message:
        lines.append(error_message)

    queries = row["queries"]
    lines.append(f"tile_count: {len(queries)}")
    for i, query in enumerate(queries):
        lines.append(f"[TILE {i}]")
        lines.append("query:")
        lines.append(query["query"])
        lines.append(f"note: {row["notes"][i]}")

        results = query["results"]
        if not results:
            continue

        lines.append("results:")
        for result in results:
            lines.append(f"path: {result['path']}")
            formatted_values =  ", ".join([format_number(value) for value in result['values']])
            lines.append(f"values: {formatted_values}")
            lines.append(f"skipped: {result['skipped']}")

    return "\n".join(lines)


def create_notebook_tool(supabase: Client, title: str, queries: list[str], notes: list[str], user_id: str) -> str:
    validate_notebook_parts(queries, notes)

    status = "working" if len(queries) > 0 else "idle"

    table = supabase.table("notebooks")
    cleaned_title = title.strip()
    inserted = table.insert({
        "title": cleaned_title,
        "queries": [{"query": query, "results": None} for query in queries],
        "notes": notes,
        "status": status,
        "error_message": None,
        "user_id": user_id
    })
    notebook_id = inserted.execute().data[0]["id"]
    return f"created notebook id={notebook_id}"


def update_notebook_tool(supabase: Client, notebook_id: int, title: str | None, queries: list[str] | None, notes: list[str] | None, user_id: str) -> str:
    notebook = fetch_notebook_row(supabase, notebook_id, user_id)
    values: dict[str, Any] = {
        "queries": clear_query_results(notebook["queries"]),
        "status": "working",
        "error_message": None,
        "last_updated": "now"
    }

    if title is not None:
        values["title"] = title.strip()

    if queries is not None:
        if notes is None:
            raise ValueError("notes must be provided when queries are replaced")

        validate_notebook_parts(queries, notes)
        values["queries"] = [{"query": query, "results": None} for query in queries]
        values["notes"] = notes
    elif notes is not None:
        validate_notebook_parts(notebook["queries"], notes)
        values["notes"] = notes

    table = supabase.table("notebooks")
    updated = table.update(values)
    owned = updated.eq("user_id", user_id)
    filtered = owned.eq("id", notebook_id)
    filtered.execute()
    return f"updated notebook id={notebook_id}"

def delete_notebook_tool(supabase: Client, notebook_id: int, user_id: str) -> str:
    fetch_notebook_row(supabase, notebook_id, user_id)
    table = supabase.table("notebooks")
    deleted = table.delete()
    owned = deleted.eq("user_id", user_id)
    owned.eq("id", notebook_id).execute()
    return f"deleted notebook id={notebook_id}"

File: crates/alphchemy_analysis/src/tools/notebook_tools.rs
use rust_supabase_sdk::SupabaseClient;
use serde::{Deserialize, Serialize};
use serde_json::{Value, json, to_value};

use crate::format::format_value;

#[derive(Debug, Deserialize)]
struct IdRow {
    id: u64
}

#[derive(Debug, Deserialize)]
struct NotebookListRow {
    id: u64,
    title: String
}

#[derive(Debug, Deserialize)]
struct NotebookRow {
    id: u64,
    last_updated: String,
    title: String,
    queries: Vec<Value>,
    notes: Vec<String>,
    status: String,
    error_message: Option<String>
}

async fn notebook_row(supabase: &SupabaseClient, notebook_id: usize, user_id: &str) -> Result<NotebookRow, String> {
    let columns = "id, last_updated, title, queries, notes, status, error_message";
    let query = supabase.from("notebooks");
    let query = query.select(columns);
    let query = query.eq("user_id", user_id);
    let query = query.eq("id", notebook_id);
    let query = query.returns::<NotebookRow>().maybe_single().execute().await;
    let row = query.map_err(|error| error.to_string())?;
    row.ok_or_else(|| format!("notebook id={notebook_id} not found"))
}

pub async fn list_notebooks(supabase: &SupabaseClient, user_id: &str) -> Result<String, String> {
    let query = supabase.from("notebooks");
    let query = query.select("id, last_updated, title");
    let query = query.eq("user_id", user_id);
    let query = query.order("last_updated", false);
    let query = query.returns::<NotebookListRow>().execute().await;
    let rows = query.map_err(|error| error.to_string())?;
    let mut lines = vec![format!("[NOTEBOOKS] {} notebook(s)", rows.len())];
    for row in rows {
        lines.push(format!("id={} title={}", row.id, row.title));
    }
    Ok(lines.join("\n"))
}

fn format_notebook(row: NotebookRow) -> String{
    let mut lines = vec![format!("id: {}", row.id), format!("last_updated: {}", row.last_updated), format!("title: {}", row.title), format!("status: {}", row.status)];
    if let Some(message) = row.error_message.filter(|message| !message.is_empty()) {
        lines.push(message);
    }
    lines.push(format!("tile_count: {}", row.queries.len()));
    for (i, query) in row.queries.iter().enumerate() {
        lines.push(format!("[TILE {i}]"));
        lines.push("query:".to_string());
        lines.push(query["query"].as_str().unwrap_or_default().to_string());
        lines.push(format!("note: {}", row.notes[i]));
        let Some(results) = query["results"].as_array() else {
            continue;
        };
        if results.is_empty() {
            continue;
        }
        lines.push("results:".to_string());
        for result in results {
            lines.push(format!("path: {}", result["path"].as_str().unwrap_or_default()));
            let values = result["values"].as_array().cloned().unwrap_or_default();
            let formatted = values.iter().map(format_value).collect::<Vec<_>>().join(", ");
            lines.push(format!("values: {formatted}"));
            lines.push(format!("skipped: {}", format_value(&result["skipped"])));
        }
    }
    lines.join("\n")
}

pub async fn view_notebook(supabase: &SupabaseClient, notebook_id: usize, user_id: &str) -> Result<String, String> {
    let row = notebook_row(supabase, notebook_id, user_id).await?;
    Ok(format_notebook(row))
}

fn validate_notebook_parts(queries: &[impl Serialize], notes: &[String]) -> Result<(), String> {
    if queries.len() != notes.len() {
        return Err("queries and notes must have the same length".to_string());
    }
    Ok(())
}

pub async fn create_notebook(supabase: &SupabaseClient, title: &str, queries: &[String], notes: &[String], user_id: &str) -> Result<String, String> {
    validate_notebook_parts(queries, notes)?;
    let status = if queries.is_empty() { "idle" } else { "working" };
    let queries = queries.iter().map(|query| json!({"query": query, "results": null})).collect::<Vec<_>>();
    let body = json!({
        "title": title.trim(),
        "queries": queries,
        "notes": notes,
        "status": status,
        "error_message": null,
        "user_id": user_id
    });
    let query = supabase.from("notebooks");
    let query = query.insert(body);
    let query = query.select_returning("id");
    let query = query.returns::<IdRow>().single().execute().await;
    let row = query.map_err(|error| error.to_string())?;
    Ok(format!("created notebook id={}", row.id))
}

pub async fn update_notebook(supabase: &SupabaseClient, notebook_id: usize, title: Option<&str>, queries: Option<&[String]>, notes: Option<&[String]>, user_id: &str) -> Result<String, String> {
    let notebook = notebook_row(supabase, notebook_id, user_id).await?;
    let cleared_queries = notebook.queries.into_iter().map(|mut query| {
        query["results"] = Value::Null;
        query
    }).collect::<Vec<_>>();
    let mut values = json!({
        "queries": cleared_queries,
        "status": "working",
        "error_message": null,
        "last_updated": "now"
    });
    if let Some(title) = title {
        values["title"] = Value::from(title.trim());
    }
    if let Some(queries) = queries {
        let Some(notes) = notes else {
            return Err("notes must be provided when queries are replaced".to_string());
        };
        validate_notebook_parts(queries, notes)?;
        values["queries"] = Value::Array(queries.iter().map(|query| json!({"query": query, "results": null})).collect());
        values["notes"] = to_value(notes).map_err(|error| error.to_string())?;
    } else if let Some(notes) = notes {
        validate_notebook_parts(&notebook.notes, notes)?;
        values["notes"] = to_value(notes).map_err(|error| error.to_string())?;
    }
    let query = supabase.from("notebooks");
    let query = query.update(values);
    let query = query.eq("user_id", user_id);
    let query = query.eq("id", notebook_id).execute().await;
    query.map_err(|error| error.to_string())?;
    Ok(format!("updated notebook id={notebook_id}"))
}

pub async fn delete_notebook(supabase: &SupabaseClient, notebook_id: usize, user_id: &str) -> Result<String, String> {
    notebook_row(supabase, notebook_id, user_id).await?;
    let query = supabase.from("notebooks");
    let query = query.delete();
    let query = query.eq("user_id", user_id);
    let query = query.eq("id", notebook_id).execute().await;
    query.map_err(|error| error.to_string())?;
    Ok(format!("deleted notebook id={notebook_id}"))
}
